### ISSUE TITLE
timestamps problem

### ISSUE THREAD

**Issue by fundon** [Automattic/mongoose#6381]
**What is the current behavior?**

```js
const mongoose = require('mongoose')
const { Schema } = mongoose

mongoose.connect('mongodb://localhost/test')

const CatSchema = new Schema(
  {
    name: String,
    age: Number,
    createdAt: {
      type: Date,
      default() {
        const d = new Date()
        d.setHours(0)
        d.setMinutes(0)
        d.setSeconds(0)
        d.setMilliseconds(0)
        return d
      }
    }
  },
  {
    timestamps: {
      updatedAt: true
    }
  }
)

const Cat = mongoose.model('Cat', CatSchema)

async function run() {
  const d = new Date()
  d.setHours(0)
  d.setMinutes(0)
  d.setSeconds(0)
  d.setMilliseconds(0)
  d.setDate(d.getDate() - 1)

  await new Cat({ name: 'ob' }).save()
  await Cat.findOneAndUpdate(
    { name: 'Zildjian' },
    {},
    { new: true, upsert: true }
  )
  await Cat.insertMany([{ name: 'a' }, { name: 'b', createdAt: d }])
  await Cat.findOneAndUpdate(
    {
      name: 'a'
    },
    {
      $set: {
        age: 2
      }
    },
    {
      new: true,
      upsert: true
    }
  )
}

run()
  .then(process.exit)
  .catch(console.error)
```

In Mongodb:

```
> db.cats.drop()
true
> db.cats.find().pretty()
{
        "_id" : ObjectId("5adcd1635abadda24b9bfbfe"),
        "name" : "ob",
        "createdAt" : ISODate("2018-04-22T16:00:00Z"),
        "updatedAt" : ISODate("2018-04-22T16:00:00Z"),
        "__v" : 0
}
{
        "_id" : ObjectId("5adcd16334c40029c9612115"),
        "name" : "Zildjian",
        "updatedAt" : ISODate("2018-04-22T18:16:03.364Z"),
        "__v" : 0,
        "createdAt" : ISODate("2018-04-22T18:16:03.364Z")
}
{
        "_id" : ObjectId("5adcd1635abadda24b9bfbff"),
        "name" : "a",
        "createdAt" : ISODate("2018-04-22T16:00:00Z"),
        "__v" : 0,
        "updatedAt" : ISODate("2018-04-22T18:16:03.372Z"),
        "age" : 2
}
{
        "_id" : ObjectId("5adcd1635abadda24b9bfc00"),
        "name" : "b",
        "createdAt" : ISODate("2018-04-21T16:00:00Z"),
        "__v" : 0,
        "updatedAt" : ISODate("2018-04-22T18:16:03.371Z")
}
>
```

Why Zildjian and ob's createdAt or updatedAt  is different?
Which behavior is right?

* node: v8.11.1
* mongodb: stable 3.6.4
* mongoose: 5.0.15

**Comment by fundon** [comment]
ping @vkarpov15

**Comment by vkarpov15** [comment]
Thanks for reporting, will investigate ASAP :+1:

**Comment by vkarpov15** [comment]
Much of the issue is due to the fact that, if you already have a `createdAt` property, mongoose will use it for timestamps unless you explicitly tell mongoose not to. You also identified a minor bug with timestamps and `createdAt`, which will be fixed with 5.0.17. With the below schema:

```javascript
const CatSchema = new Schema(
  {
    name: String,
    age: Number,
    createdAt: {
      type: Date,
      default: function() {
        const d = new Date()
        d.setHours(0)
        d.setMinutes(0)
        d.setSeconds(0)
        d.setMilliseconds(0)
        return d
      }
    }
  },
  {
    timestamps: {
      createdAt: false,
      updatedAt: true
    }
  }
)
```

You'll get the below output with mongoose 5.0.17:

```
$ node gh-6381.js 
Res [ { createdAt: 2018-04-27T04:00:00.000Z,
    _id: 5ae33592ea592930a9d3b746,
    name: 'ob',
    updatedAt: 2018-04-27T14:37:06.602Z,
    __v: 0 },
  { createdAt: 2018-04-27T04:00:00.000Z,
    _id: 5ae335926ed062c0fce7998f,
    name: 'Zildjian',
    updatedAt: 2018-04-27T14:37:06.608Z,
    __v: 0 },
  { createdAt: 2018-04-27T04:00:00.000Z,
    _id: 5ae33592ea592930a9d3b747,
    name: 'a',
    __v: 0,
    updatedAt: 2018-04-27T14:37:06.616Z,
    age: 2 },
  { createdAt: 2018-04-26T04:00:00.000Z,
    _id: 5ae33592ea592930a9d3b748,
    name: 'b',
    __v: 0,
    updatedAt: 2018-04-27T14:37:06.615Z } ]
$ 
```
